python/torch_mlir/eager/torch_dispatch.py

```diff
@@ -0,0 +1,353 @@
+                            normalize_to_only_use_kwargs,
+                        )
+                        return (
+                            candidate_schema,
+                            new_args_and_kwargs.args,
+                            new_args_and_kwargs.kwargs,
+                        )
+            else:
+                # matched more than one schema despite providing types. Not sure if this is possible since in the reference
+                # implementation the comment here states caller should provide types (optional for the reference impl)
+                # where we require arg types.
+                schema_printouts = "\n".join(str(schema) for schema in matches)
+                raise RuntimeError(
+                    f"Tried to normalize arguments to {torch.typename(target)} but "
+                    f"the schema match was ambiguous! Please provide argument types to "
+                    f"the normalize_arguments() call. Available schemas:\n{schema_printouts}"
+                )
+
+    raise RuntimeError("couldn't normalize args")
+
+
+# shameless SO copy pasta
+def urlify(s):
+    # Remove all non-word characters (everything except numbers and letters)
+    s = re.sub(r"[^\w\s]", "", s)
+    # Replace all runs of whitespace with a single dash
+    s = re.sub(r"\s+", "_", s)
+
+    return s
+
+
+def build_script_function(
+    schema: torch._C.FunctionSchema,
+    args: List[torch._C.Argument],
+    kwargs: Dict[str, Any],
+) -> torch.jit.ScriptFunction:
+    """
+    Build a torch.jit.ScriptFunction that corresponds to the schema.
+    This is necessary so that we can pass a "hydrated" module to ModuleBuilder::importModule.
+    Hydrated meaning with inlined constants with correct values; a gotcha here is that if you don't inline the constants
+    then importModule assume default values (i think?) thereby (potentially) failing to support the op (not sure how this makes sense but native_batch_norm for example doesn't work without train=False)
+
+    Works by
+    1. creating an empty ts graph
+    2. creating and inserting a node corresponding to the op; nodes inserted are just names (hence following steps)
+    2. for each arg to the op:
+        a. if arg is a tensor: add input to the graph corresponding to arg
+           if arg is a constant (bool, int, list): inline a constant (at the top of the graph)
+        b. wire value from previous step to the op (ie SSA value corresponding to graph input is "passed" to op)
+        c. if arg is a tensor: fill in shape
+
+    The (poorly handled/poorly understood) gotcha here is args with type Optional[Tensors] with values None.
+    These pytorch types correspond to !torch.optional<!torch.vtensor>, which torch-mlir complains about having a torch.type_bound
+    (something something torch.type_bound can only be a func arg_attr for torch.vtensor). The hacky solution here works but
+    there's probably some straightforward way to handle this that I didn't stumble upon.
```

Comment:
This is a place where I expect someone to tell me that there's a more appropriate way to handle it.

Reply:
Right now we don't have a calling convention that can pass optional types dynamically. So the optionality should be part of the compiled cache key (that is, we specialize for either the None or not-None case).